velox/expression/SimpleFunctionAdapter.h

```diff
@@ -313,21 +317,91 @@ class SimpleFunctionAdapter : public VectorFunction {
       unpack<POSITION + 1>(
           applyContext, nextNonNull, packed, readers..., oneReader);
     } else {
-      auto* oneUnpacked = packed.at(POSITION);
-      auto oneReader = VectorReader<arg_at<POSITION>>(oneUnpacked);
+      if constexpr (
+          std::is_same_v<arg_at<POSITION>, int8_t> ||
+          std::is_same_v<arg_at<POSITION>, int16_t> ||
+          std::is_same_v<arg_at<POSITION>, int32_t> ||
+          std::is_same_v<arg_at<POSITION>, int64_t> ||
+          std::is_same_v<arg_at<POSITION>, float> ||
+          std::is_same_v<arg_at<POSITION>, double>) {
+        if (packed[POSITION]->encoding() == VectorEncoding::Simple::FLAT) {
+          auto& flatUnpacked = packed[POSITION];
+          auto oneReader =
```

Comment:
Would it be possible to move this to a helper function in DecodedArgs, which based on encoding returns the best VectorReader, so FlatVectorReader if flat, ConstVectorReader if const. 
Unless I am mistaken moving this logic there will reduce code duplication.